zvariant/src/file_path.rs

```diff
@@ -0,0 +1,293 @@
+/// bellow for reference
+/// ```
+///    use zvariant::FilePath;
+///    use std::path::{Path, PathBuf};
+///
+///    let path = Path::new("/hello/world\0");
+///    let path_buf = PathBuf::from(path);
+///
+///    let p1 = FilePath::from(path);
+///    let p2 = FilePath::from(path_buf);
+///    let p3 = FilePath::from("/hello/world");
+///
+///
+///    assert_eq!(p1, p2);
+///    assert_eq!(p2, p3);
+/// ```
+#[derive(Type, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
+#[zvariant(signature = "ay")]
+pub struct FilePath<'f>(Cow<'f, CStr>);
+
+impl<'f> FilePath<'f> {
+    pub fn new(cow: Cow<'f, CStr>) -> Self {
+        Self(cow)
+    }
+}
+
+impl From<CString> for FilePath<'_> {
+    fn from(value: CString) -> Self {
+        FilePath(Cow::Owned(value))
+    }
+}
+
+impl<'f> From<&'f CString> for FilePath<'f> {
+    fn from(value: &'f CString) -> Self {
+        FilePath(Cow::Borrowed(value.as_c_str()))
+    }
+}
+
+impl<'f> From<&'f OsStr> for FilePath<'f> {
+    fn from(value: &'f OsStr) -> FilePath<'f> {
+        FilePath(bytes_with_null(value.as_encoded_bytes()))
+    }
+}
+
+impl<'f> From<&'f OsString> for FilePath<'f> {
+    fn from(value: &'f OsString) -> FilePath<'f> {
+        FilePath(bytes_with_null(value.as_encoded_bytes()))
+    }
+}
+
+impl From<OsString> for FilePath<'_> {
+    fn from(value: OsString) -> Self {
+        FilePath(vec_to_cstr(value.as_encoded_bytes().to_vec()))
+    }
+}
+
+impl<'f> From<&'f PathBuf> for FilePath<'f> {
+    fn from(value: &'f PathBuf) -> FilePath<'f> {
+        FilePath::from(value.as_os_str())
+    }
+}
+
+impl<'f> From<PathBuf> for FilePath<'f> {
+    fn from(value: PathBuf) -> FilePath<'f> {
+        FilePath::from(OsString::from(value))
+    }
+}
+
+impl<'f> From<&'f Path> for FilePath<'f> {
+    fn from(value: &'f Path) -> Self {
+        Self::from(value.as_os_str())
+    }
+}
+
+impl<'f> From<&'f CStr> for FilePath<'f> {
+    fn from(value: &'f CStr) -> Self {
+        Self(Cow::Borrowed(value))
+    }
+}
+
+impl<'f> From<&'f str> for FilePath<'f> {
+    fn from(value: &'f str) -> Self {
+        Self::from(OsStr::new(value))
+    }
+}
+
+impl<'f> AsRef<FilePath<'f>> for FilePath<'f> {
+    fn as_ref(&self) -> &FilePath<'f> {
+        self
+    }
+}
+
+impl<'f> From<FilePath<'f>> for OsString {
+    fn from(value: FilePath<'f>) -> Self {
+        unsafe { OsString::from_encoded_bytes_unchecked(value.0.to_bytes().to_vec()) }
+    }
+}
+
+impl<'f> From<&'f FilePath<'f>> for &'f Path {
+    fn from(value: &'f FilePath<'f>) -> Self {
+        // unwrap?
+        Path::new(value.0.as_ref().to_str().unwrap())
+    }
+}
+
+impl<'f> From<FilePath<'f>> for PathBuf {
+    fn from(value: FilePath<'f>) -> Self {
+        PathBuf::from(value.0.to_string_lossy().to_string())
+    }
+}
+
+/// Converts a `Vec<u8>` into a null-terminated `CStr`.
+///
+/// Truncates the vector at the first null byte, if present.
+/// If no null byte exists, appends one to ensure proper termination.
```

Comment:
I really appreciate you adding proper docs for even the internal API. :+1: This is why I hate to nitpick here but please don't break lines for starting a new sentence (only for paragraphs). Break them on word boundary at 100 chars, just like the code. :pray: Same goes for other comments.